File: src/content/docs/topics/cs/coding-problems/backtracking/039-combination-sum-approach2.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

function combinationSum(candidates: number[], target: number): number[][] {
    const result: number[][] = [];
    const path: number[] = [];

    function backtrack(start: number, remaining: number): void {
        if (remaining === 0) {
            result.push([...path]);     // L1: O(k) copy at leaf
            return;
        }
        if (remaining < 0) return;
        for (let i = start; i < candidates.length; i++) {
            path.push(candidates[i]);                       // L2: O(1) push
            backtrack(i, remaining - candidates[i]);        // L3: recurse (reuse allowed)
            path.pop();                                     // L4: O(1) pop
        }
    }

    backtrack(0, target);
    return result;
}

const norm = (arr: number[][]): string =>
    JSON.stringify(arr.map(a => [...a].sort((x, y) => x - y)).sort((a, b) => JSON.stringify(a) < JSON.stringify(b) ? -1 : 1));

const r = combinationSum([2, 3, 6, 7], 7);
assert(norm(r) === norm([[2, 2, 3], [7]]));
const r2 = combinationSum([2, 3, 5], 8);
assert(norm(r2) === norm([[2, 2, 2, 2], [2, 3, 3], [3, 5]]));
assert(norm(combinationSum([3], 9)) === norm([[3, 3, 3]]));
assert(combinationSum([5], 3).length === 0);
console.log('all tests pass');

File: src/content/docs/topics/cs/coding-problems/backtracking/039-combination-sum-approach2.swift
// LEETCODE_TYPE: Solution
func expectEqual<T: Equatable>(
    _ actual: T,
    _ expected: T,
    _ message: String = "",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard actual == expected else {
        let detail = message.isEmpty ? "values differ" : message
        fatalError("\(file):\(line): \(detail). Expected \(expected), got \(actual)")
    }
}

func expectTrue(
    _ condition: @autoclosure () -> Bool,
    _ message: String = "expected true",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard condition() else {
        fatalError("\(file):\(line): \(message)")
    }
}

func reportSuccess() {
    print("All Swift tests passed")
}

final class Solution {
    private func normalized(_ values: [[Int]]) -> [[Int]] {
        values.sorted { $0.lexicographicallyPrecedes($1) }
    }

    func combinationSum(_ candidates: [Int], _ target: Int) -> [[Int]] {
        let values = candidates.sorted(); var result: [[Int]] = []
        func search(_ start: Int, _ remaining: Int, _ current: [Int]) { if remaining == 0 { result.append(current); return }; for index in start..<values.count where values[index] <= remaining { search(index, remaining - values[index], current + [values[index]]) } }
        search(0, target, []); return normalized(result)
    }
}

func runTests() {
    // TEST_VECTORS_BEGIN sha256:b930946d881468ede6b88d809c32a58b410fd6644cdc0f4908dedfa836c7006c
    expectEqual(Solution().combinationSum([2, 3, 6, 7], 7), [[2, 2, 3], [7]], "canonical")
    expectEqual(Solution().combinationSum([2, 3, 5], 8), [[2, 2, 2, 2], [2, 3, 3], [3, 5]], "two-solutions")
    expectEqual(Solution().combinationSum([2], 1), [], "none")
    expectEqual(Solution().combinationSum([1], 2), [[1, 1]], "single-repeat")
    expectEqual(Solution().combinationSum([5, 6], 5), [[5]], "exact")
    // EXCLUDED_VECTOR nonpositive: [[0,1],1] | Candidates are distinct positive integers.
    // TEST_VECTORS_END
    reportSuccess()
}

runTests()
